### PR TITLE
Face objdetect

### PR DESCRIPTION
**Pull request by abdelaziz-mahdy**
fixing this issue https://github.com/rainyl/opencv_dart/issues/88 

i dont know how to implement the tests for the classes, so let me know if you follow example or something i can follow

### CODE REVIEW

**Comment by rainyl** [comment]
Great! I will review it later.

**Comment by rainyl** [comment]
@abdelaziz-mahdy I have added a simple test for `cv.FaceDetectorYN`, referring to https://docs.opencv.org/4.x/d0/dd4/tutorial_dnn_face.html and https://github.com/opencv/opencv/blob/4.x/samples/dnn/face_detect.cpp , you can add more test according to the tutorial of opencv.

It works now.

image

**Comment by abdelaziz-mahdy** [comment]
> @abdelaziz-mahdy I have added a simple test for `cv.FaceDetectorYN`, referring to https://docs.opencv.org/4.x/d0/dd4/tutorial_dnn_face.html and https://github.com/opencv/opencv/blob/4.x/samples/dnn/face_detect.cpp , you can add more test according to the tutorial of opencv.
> 
> It works now.
> 
> image

the model "test/models/face_detection_yunet_2023mar.onnx" is missing i guess this is why the tests fail, anyway will add another test for the other class too

**Comment by rainyl** [comment]
yes, but just ignore it now, you can find the link of the models in the above link, I will add the models when finished.

**Comment by abdelaziz-mahdy** [comment]
> yes, but just ignore it now, you can find the link of the models in the above link, I will add the models when finished.

when running locally Invalid argument(s): Failed to load dynamic library 'libopencv_dart.dylib': dlopen(libopencv_dart.dylib, 0x0001): tried: 'libopencv_dart.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibopencv_dart.dylib' (no such file), '/Users/AbdelazizMahdy/Developer/flutter/bin/cache/artifacts/engine/darwin-x64/./libopencv_dart.dylib' (no such file), '/usr/local/lib/./libopencv_dart.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/./libopencv_dart.dylib' (no such file), '/Users/AbdelazizMahdy/Developer/flutter/bin/cache/artifacts/engine/darwin-x64/../../../libopencv_dart.dylib' (no such file), '/Users/AbdelazizMahdy/Developer/flutter/bin/cache/artifacts/engine/darwin-x64/Frameworks/libopencv_dart.dylib' (no such file), '/Users/AbdelazizMahdy/Developer/flutter/bin/cache/artifacts/engine/darwin-x64/./libopencv_dart.dylib' (no such file), '/usr/local/lib/./libopencv_dart.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/./libopencv_dart.dylib' (no such file), '/Users/AbdelazizMahdy/Developer/flutter/bin/cache/artifacts/engine/darwin-x64/../../../libopencv_dart.dylib' (no such file), '/Users/AbdelazizMahdy/Developer/flutter/bin/cache/artifacts/engine/darwin-x64/Frameworks/libopencv_dart.dylib' (no such file), '/usr/lib/libopencv_dart.dylib' (no such file, not in dyld cache), 'libopencv_dart.dylib' (no such file)
dart:ffi                                                 new DynamicLibrary.open
package:opencv_dart/src/core/base.dart 43:31             loadNativeLibrary
package:opencv_dart/src/core/base.dart 51:23             CFFI
package:opencv_dart/src/core/base.dart                   CFFI
package:opencv_dart/src/objdetect/objdetect.dart 630:15  new FaceDetectorYN.fromFile.<fn>.<fn>
package:opencv_dart/src/core/base.dart 67:22             cvRun
package:opencv_dart/src/objdetect/objdetect.dart 629:7   new FaceDetectorYN.fromFile.<fn>
package:ffi/src/arena.dart 124:31                        using
package:opencv_dart/src/objdetect/objdetect.dart 626:12  new FaceDetectorYN.fromFile
test/objdetect_test.dart 212:40                          main.<fn>
i was going to depend on ci :)

**Comment by rainyl** [comment]
https://github.com/rainyl/opencv_dart/blob/ee4f263165501347e625e6e9fd4fdf37ffc782b4/.github/workflows/build_test_release.yaml#L246

You need to add the path of libopencv_dart.dylib to DYLD_LIBRARY_PATH, 
```bash
export DYLD_LIBRARY_PATH=`pwd`/macos:$DYLD_LIBRARY_PATH
```

or add the path to `dart.env` in settings.json for vscode (if you are using vscode) to make tests side menu work.

or, you can just change DYLD_LIBRARY_PATH in ~/.zshrc

**Comment by abdelaziz-mahdy** [comment]
> https://github.com/rainyl/opencv_dart/blob/ee4f263165501347e625e6e9fd4fdf37ffc782b4/.github/workflows/build_test_release.yaml#L246
> 
> You need to add the path of libopencv_dart.dylib to DYLD_LIBRARY_PATH,
> 
> ```shell
> export DYLD_LIBRARY_PATH=`pwd`/macos:$DYLD_LIBRARY_PATH
> ```
> 
> or add the path to `dart.env` in settings.json for vscode (if you are using vscode) to make tests side menu work.
> 
> or, you can just change DYLD_LIBRARY_PATH in ~/.zshrc

yes i found those in cli, but i cant get it to work 
anyway i added a test case, logically it should work

**Comment by rainyl** [comment]
I have added an option to get `OPENCV_DART_LIB_PATH`, now you can define the variable to make it work.

```bash
export OPENCV_DART_LIB_PATH=`pwd`/macos/libopencv_dart.dylib
dart test test/objdetect_test.dart
```

File: src/objdetect/objdetect.h [hunk 1]
```diff
@@ -110,6 +114,29 @@ CvStatus QRCodeDetector_setEpsX(QRCodeDetector self, double epsX);
...
+                            FaceDetectorYN *rval);
+CvStatus FaceDetectorYN_NewFromBuffer(const char *framework, VecUChar buffer, VecUChar buffer_config,
+                                      Size input_size, float score_threshold, float nms_threshold, int top_k,
+                                      int backend_id, int target_id, FaceDetectorYN *rval);
+void     FaceDetectorYN_Close(FaceDetectorYNPtr self);
+CvStatus FaceDetectorYN_Detect(FaceDetectorYN self, Mat img, Mat *faces);
+CvStatus FaceDetectorYN_SetInputSize(FaceDetectorYN self, Size input_size);
+CvStatus FaceDetectorYN_SetScoreThreshold(FaceDetectorYN self, float score_threshold);
+CvStatus FaceDetectorYN_SetNMSThreshold(FaceDetectorYN self, float nms_threshold);
+CvStatus FaceDetectorYN_SetTopK(FaceDetectorYN self, int top_k);
```
Comment (rainyl) [inline src/objdetect/objdetect.h:129]: Here, I think we should also add getters such as `getInputSize ()`, `getNMSThreshold`, etc. https://docs.opencv.org/4.x/df/d20/classcv_1_1FaceDetectorYN.html#a68b6fb9bffbed0f3d5c104996113f247
```suggestion
CvStatus FaceDetectorYN_SetTopK(FaceDetectorYN self, int top_k);
```

Comment (abdelaziz-mahdy) [reply src/objdetect/objdetect.h:129]: Ok will do it 👍

Comment (abdelaziz-mahdy) [reply src/objdetect/objdetect.h]: done

**Comment by abdelaziz-mahdy** [comment]
i added getters and tests for them, can you let me know what is missing, since i cant figure it from the codecov

**Comment by rainyl** [comment]
> i added getters and tests for them, can you let me know what is missing, since i cant figure it from the codecov

Looks good to me, but I personally don't like the line length limitation (80 characters), so I am going to format them, then it will be merged.

Thanks for your contributions!

**Comment by abdelaziz-mahdy** [comment]
> > i added getters and tests for them, can you let me know what is missing, since i cant figure it from the codecov
> 
> Looks good to me, but I personally don't like the line length limitation (80 characters), so I am going to format them, then it will be merged.
> 
> Thanks for your contributions!

i use the default dart formatter, also i think you can set your own rules in file so it gets used when other people are working on it

**Comment by rainyl** [comment]
> > > i added getters and tests for them, can you let me know what is missing, since i cant figure it from the codecov
> > 
> > 
> > Looks good to me, but I personally don't like the line length limitation (80 characters), so I am going to format them, then it will be merged.
> > Thanks for your contributions!
> 
> i use the default dart formatter, also i think you can set your own rules in file so it gets used when other people are working on it

Thanks~ I will set the rules in the future. ❤️

**Comment by abdelaziz-mahdy** [comment]
also thank you for your help, i wouldnt have done it without it, i look forward to using the new modules in my app and see how it performs

**Review by rainyl — APPROVED** [review]
Ready

**Review by rainyl — APPROVED** [review]

**Pull request merged**